.github/workflows/gen_matrix.pl

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env swipl
+
+%% This script is used to generate an output matrix (in JSON format)
+%% that the github actions can import and use to generate various test
+%% configurations.
+%%
+%% There are three main axes: os, compiler, and solver.  The os and
+%% compiler axis have a vector of versions, one or more of which
+%% should be designated as the main version.  The solver axis has a
+%% solver version vector for each defined solver, where again at least
+%% one of each should be designated as the main solver version.
+%%
+%% Adding versions or identifiers to any axis is done with a simple
+%% declarative statement in the first section below.
+%%
+%% The output matrix defines a more efficient selection of tests than
+%% a full cartesian product (because the latter would uselessly test
+%% many redundant configurations).
+%%
+%%  1. Test the main version of each solver on the code build with the
+%%     main version of the compiler, for each operating system
+%%     (solver_main_compiler_main).
+%%
+%%  2. Test the other versions of compiled code with the main version
+%%     of each solver, for each operating system
+%%     (solver_main_compiler_all).
+%%
+%%  3. Test the other versions of each solver on the main compiler
+%%     version of the code, for each operating system
+%%     (solver_all_compiler_main).  Note that the tests run against
+%%     each solver independently, but a test run will test all
+%%     available solvers (and skip any unavailable solvers), so the
+%%     most efficient matrix varies each solver version independently
+%%     (but in parallel) for each test run until all solver versions
+%%     are exhausted.
+%%
+%% This script can be run by hand to see the generated JSON output
+%% using SWI Prolog:
+%%
+%%    $ swipl .github/workflows/gen_matrix.pl
+
+:- initialization(main, main).
+:- use_module(library(http/json)).
+:- use_module(library(clpfd)).
+
+%% Declare every Operating System that should be tested (separately in
+%% sequence)
+os(ubuntu).
+
+%% Declare every Compiler that should be used to build the code
+%% (separately in sequence)
+compiler(ghc).
+
+%% Declare every Solver that should be used (in parallel, not in
+%% sequence: one version of every solver is chosen for each test.
+solver(z3).
+solver(yices).
+solver(stp).
+solver(cvc4).
+solver(boolector).
+solver(abc).
+
+%% Specify the main version of each variable (i.e. the most interesting value of that variable.
+main_version(ubuntu, "ubuntu-latest").
+main_version(ghc, "8.10.7").
+main_version(z3, "4_8_14").
+main_version(yices, "2_6").
+main_version(stp, "2_3_3").
+main_version(cvc4, "1_8").
+main_version(boolector, "3_2_2").
+main_version(abc, "2021_12_30").
+
+%% Specify the versions of the operating system(s), along with the
+%% main version(s) that are of primary interest.
+version(ubuntu, ubuntu_latest).
```

Comment:
This might be my lack of SWI Prolog experience showing, but I'm unclear why the `ubuntu` version used here (`ubuntu_latest`) is different from the corresponding version used above in the `main_version` (`"ubuntu-latest"`, a string literal with a hyphen instead of an underscore).

Reply:
Good catch!  Since we always test every configuration against all operating systems (at present) this fact was actually unused, but it would have bitten us in some future update where it became more important.